### PR TITLE
Handle Epoch Boundary Misses

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Bug Fix

**What does this PR do? Why is it needed?**

In certain situations in the network, we can miss saving the epoch boundary state into our boundary cache. This can cause issues in future epochs, as if we need to fetch the justified/finalized state we end up having to regenerating it from scratch. Instead of doing that, we have now decided to add an optimization to always save epoch boundary states into our cache. 

We also handle the edge case where the incorrect epoch boundary state is saved into our cache in the event of shallow re-orgs(1 slot) at the start of an epoch. Both these cases have test cases added in along with bug fixes in the deletion of values for the boundary cache.

**Which issues(s) does this PR fix?**

N.A

**Other notes for review**

### CODE REVIEW

File: beacon-chain/state/stategen/setter.go [hunk 1]
```diff
@@ -79,6 +81,52 @@ func (s *State) saveStateByRoot(ctx context.Context, blockRoot [32]byte, st stat
...
+			// simply skip this step.
+			if s.hotStateCache.has([32]byte(bRoot)) {
+				log.WithFields(logrus.Fields{
+					"slot": epochStart,
+					"root": fmt.Sprintf("%#x", bRoot),
+				}).Debug("Recovering state for epoch boundary cache")
+
+				// Remove the non-canonical state from our cache.
+				if nonCanonicalState {
+					if err := s.epochBoundaryStateCache.delete(rInfo.root); err != nil {
```
Comment (rkapka) [inline beacon-chain/state/stategen/setter.go:120]: Shouldn't we delete the non-canonical state even if we don't have the correct state in the hot state cache?

File: beacon-chain/state/stategen/setter.go [hunk 2]
```diff
@@ -79,6 +81,52 @@ func (s *State) saveStateByRoot(ctx context.Context, blockRoot [32]byte, st stat
...
+		rInfo, ok, err := s.epochBoundaryStateCache.getBySlot(epochStart)
+		if err != nil {
+			return err
+		}
+		bRoot, err := helpers.BlockRootAtSlot(st, epochStart)
+		if err != nil {
+			return err
+		}
+
+		nonCanonicalState := ok && [32]byte(bRoot) != rInfo.root
```
Comment (terencechain) [inline beacon-chain/state/stategen/setter.go:100]: Note this doesn't help if you have an epoch worth of skip slots

File: beacon-chain/state/stategen/setter.go (hunk 1 above)
Comment (nisdas) [reply beacon-chain/state/stategen/setter.go:120]: With how the current cache is read/written to, there isn't too much harm just leaving it in there. I think we should only remove it, if the 'correct/canonical` alternative exists

File: beacon-chain/state/stategen/setter.go (hunk 2 above)
Comment (nisdas) [reply beacon-chain/state/stategen/setter.go:100]: Yeap, it can't help there, only in cases where its in the same epoch.

File: beacon-chain/state/stategen/setter.go [hunk 3]
```diff
@@ -79,6 +81,52 @@ func (s *State) saveStateByRoot(ctx context.Context, blockRoot [32]byte, st stat
...
+				// Remove the non-canonical state from our cache.
+				if nonCanonicalState {
+					if err := s.epochBoundaryStateCache.delete(rInfo.root); err != nil {
+						return err
+					}
+				}
+				hState := s.hotStateCache.get([32]byte(bRoot))
+				if err := s.epochBoundaryStateCache.put([32]byte(bRoot), hState); err != nil {
+					return err
+				}
```
[the remark this answers is not in the document]
Comment (nisdas) [reply beacon-chain/state/stategen/setter.go:127]: I want to hold off on that, one of the weaknesses of the approach in the PR, is that we do not have access to the canonical methods in forkchoice ( without some nasty recursive locks). So we do this check for every incoming block. In the event we have a non-canonical block coming in, we do not want to remove it completely from the hot-state cache. This is as the next block coming in could rewrite this particular entry in the boundary cache( and if we remove the hot state entry, that state completely disappears).

**Review by terencechain — APPROVED** [review]

**Review by terencechain — APPROVED** [review]
